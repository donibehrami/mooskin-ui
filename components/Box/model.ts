// import { IMooskinContext, IStyledTheme, IThemeBackgroundColors, IThemeBorderColors, IThemeFontColors } from '../Styled/model';
import { IMooskinContext } from '../Styled/model';

export type IntensityType = 'xs' | 'sm' | 'base' | 'md' | 'lg' | 'xl' | '2xl';
export type BoxShadowIntensityType = IntensityType | 'inner' | 'outline';
// export type NestedThemeType = [keyof IStyledTheme, keyof IThemeFontColors | keyof IThemeBorderColors | keyof IThemeBackgroundColors];
export type NestedThemeType = [string, string];

export const boxComponentProps = [
	'id',
	'className',
	'style',
	'm',
	'mt',
	'mr',
	'mb',
	'ml',
	'mx',
	'my',
	'boxShadow',
	'color',
	'round',
	'p',
	'pt',
	'pr',
	'pb',
	'pl',
	'px',
	'py',
	'opacity',
	'fontFamily',
	'fontSize',
	'fontWeight',
	'lineHeight',
	'textAlign',
	'fontStyle',
	'textTransform',
	'textDecoration',
	'w',
	'h',
	'minW',
	'maxW',
	'minH',
	'maxH',
	'd',
	'overflow',
	'overflowX',
	'overflowY',
	'align',
	'alignContent',
	'alignSelf',
	'justifyItems',
	'justify',
	'justifySelf',
	'wrap',
	'direction',
	'flex',
	'flexGrow',
	'flexShrink',
	'flexBasis',
	'order',
	'bg',
	'bgImage',
	'bgColor',
	'bgSize',
	'bgPosition',
	'bgRepeat',
	'border',
	'borderWidth',
	'borderStyle',
	'borderColor',
	'borderTop',
	'borderTopWidth',
	'borderTopStyle',
	'borderTopColor',
	'borderRight',
	'borderRightWidth',
	'borderRightStyle',
	'borderRightColor',
	'borderBottom',
	'borderBottomWidth',
	'borderBottomStyle',
	'borderBottomColor',
	'borderLeft',
	'borderLeftWidth',
	'borderLeftStyle',
	'borderLeftColor',
	'borderRadius',
	'borderTopLeftRadius',
	'borderTopRightRadius',
	'borderBottomRightRadius',
	'borderBottomLeftRadius',
	'position',
	'zIndex',
	'top',
	'right',
	'bottom',
	'left',
	'animation',
	'visibility',
	'cursor',
	'boxAs',
	'children',
	'theme',
];

export type IBoxComponentProps = IBaseBoxComponentProps & React.HTMLAttributes<HTMLElement>;
export type IBoxImageComponentProps = IBaseBoxComponentProps & React.ImgHTMLAttributes<HTMLImageElement>;
export type IBoxLabelComponentProps = IBaseBoxComponentProps & React.LabelHTMLAttributes<HTMLLabelElement>;
export type IButtonBoxComponentProps = IBoxComponentProps & React.ButtonHTMLAttributes<HTMLButtonElement>;
export type IInputBoxComponentProps = IBoxComponentProps & React.InputHTMLAttributes<HTMLInputElement>;
export type IAnchorBoxComponentProps = IBoxComponentProps & React.AnchorHTMLAttributes<HTMLAnchorElement>;
export type ITextAreaBoxComponentProps = IBoxComponentProps & React.TextareaHTMLAttributes<HTMLTextAreaElement>;

export interface IBaseBoxComponentProps extends IMooskinContext {
	ref?: any;

	/** render Box as a different html element */
	boxAs?: any;

	/** margin */
	m?: React.CSSProperties['margin'];

	/** margin top */
	mt?: React.CSSProperties['marginTop'];

	/** margin right */
	mr?: React.CSSProperties['marginRight'];

	/** margin bottom */
	mb?: React.CSSProperties['marginBottom'];

	/** margin left */
	ml?: React.CSSProperties['marginLeft'];

	/** margin left & margin right */
	mx?: number | string;

	/** margin top & margin-bottom */
	my?: number | string;

	/** padding */
	p?: React.CSSProperties['padding'];

	/** padding top */
	pt?: React.CSSProperties['paddingTop'];

	/** padding right */
	pr?: React.CSSProperties['paddingRight'];

	/** padding bottom */
	pb?: React.CSSProperties['paddingBottom'];

	/** padding left */
	pl?: React.CSSProperties['paddingLeft'];

	/** padding left & padding right */
	px?: number | string;

	/** padding top & padding-bottom */
	py?: number | string;

	/** color */
	color?: React.CSSProperties['color'] | NestedThemeType | any;

	/** opacity */
	opacity?: React.CSSProperties['opacity'];

	/** font family */
	fontFamily?: React.CSSProperties['fontFamily'];

	/** font size */
	fontSize?: React.CSSProperties['fontSize'];

	/** font weight */
	fontWeight?: React.CSSProperties['fontWeight'];

	/** line height */
	lineHeight?: React.CSSProperties['lineHeight'];

	/** text align */
	textAlign?: React.CSSProperties['textAlign'];

	/** font style */
	fontStyle?: React.CSSProperties['fontStyle'];

	/** text transform */
	textTransform?: React.CSSProperties['textTransform'];

	/** text decoration */
	textDecoration?: React.CSSProperties['textDecoration'];

	/** width */
	w?: React.CSSProperties['width'];

	/** height */
	h?: React.CSSProperties['height'];

	/** max width */
	maxW?: React.CSSProperties['maxWidth'];

	/** min width */
	minW?: React.CSSProperties['minWidth'];

	/** max height */
	maxH?: React.CSSProperties['maxHeight'];

	/** min height */
	minH?: React.CSSProperties['minHeight'];

	/** display */
	d?: React.CSSProperties['display'];

	/** overflow */
	overflow?: React.CSSProperties['overflow'];

	/** overflow X */
	overflowX?: React.CSSProperties['overflowX'];

	/** overflow Y */
	overflowY?: React.CSSProperties['overflowY'];

	/** align items */
	align?: React.CSSProperties['alignItems'];

	/** align content */
	alignContent?: React.CSSProperties['alignContent'];

	/** align self */
	alignSelf?: React.CSSProperties['alignSelf'];

	/** justify items */
	justifyItems?: React.CSSProperties['justifyItems'];

	/** justify content */
	justify?: React.CSSProperties['justifyContent'];

	/** justify self */
	justifySelf?: React.CSSProperties['justifySelf'];

	/** flex wrap */
	flexWrap?: React.CSSProperties['flexWrap'];

	/** flex direction */
	direction?: React.CSSProperties['flexDirection'];

	/** flex */
	flex?: React.CSSProperties['flex'];

	/** flex grow */
	flexGrow?: React.CSSProperties['flexGrow'];

	/** flex shrink */
	flexShrink?: React.CSSProperties['flexShrink'];

	/** flex basis */
	flexBasis?: React.CSSProperties['flexBasis'];

	/** order */
	order?: React.CSSProperties['order'];

	/** background */
	bg?: React.CSSProperties['background'];

	/** background image */
	bgImage?: React.CSSProperties['backgroundImage'];

	/** background color */
	bgColor?: React.CSSProperties['backgroundColor'] | NestedThemeType;

	/** background size */
	bgSize?: React.CSSProperties['backgroundSize'];

	/** background position */
	bgPosition?: React.CSSProperties['backgroundPosition'];

	/** background repeat */
	bgRepeat?: React.CSSProperties['backgroundRepeat'];

	/** border */
	border?: React.CSSProperties['border'];

	/** border width */
	borderWidth?: React.CSSProperties['borderWidth'];

	/** border style */
	borderStyle?: React.CSSProperties['borderStyle'];

	/** border color */
	borderColor?: React.CSSProperties['borderColor'] | NestedThemeType;

	/** border top */
	borderTop?: React.CSSProperties['borderTop'];

	/** border top width */
	borderTopWidth?: React.CSSProperties['borderTopWidth'];

	/** border top style */
	borderTopStyle?: React.CSSProperties['borderTopStyle'];

	/** border top color */
	borderTopColor?: React.CSSProperties['borderTopColor'];

	/** border right */
	borderRight?: React.CSSProperties['borderRight'];

	/** border right width */
	borderRightWidth?: React.CSSProperties['borderRightWidth'];

	/** border right style */
	borderRightStyle?: React.CSSProperties['borderRightStyle'];

	/** border right color */
	borderRightColor?: React.CSSProperties['borderRightColor'];

	/** border bottom */
	borderBottom?: React.CSSProperties['borderBottom'];

	/** border bottom width */
	borderBottomWidth?: React.CSSProperties['borderBottomWidth'];

	/** border bottom style */
	borderBottomStyle?: React.CSSProperties['borderBottomStyle'];

	/** border bottom color */
	borderBottomColor?: React.CSSProperties['borderBottomColor'];

	/** border left */
	borderLeft?: React.CSSProperties['borderLeft'];

	/** border left width */
	borderLeftWidth?: React.CSSProperties['borderLeftWidth'];

	/** border left style */
	borderLeftStyle?: React.CSSProperties['borderLeftStyle'];

	/** border left color */
	borderLeftColor?: React.CSSProperties['borderLeftColor'];

	/** border radius */
	borderRadius?: React.CSSProperties['borderRadius'];

	/** border radius top left */
	borderTopLeftRadius?: React.CSSProperties['borderTopLeftRadius'];

	/** border radius top right */
	borderTopRightRadius?: React.CSSProperties['borderTopRightRadius'];

	/** border radius bottom right */
	borderBottomRightRadius?: React.CSSProperties['borderBottomRightRadius'];

	/** border radius bottom left */
	borderBottomLeftRadius?: React.CSSProperties['borderBottomLeftRadius'];

	/** position */
	position?: React.CSSProperties['position'];

	/** z index */
	zIndex?: React.CSSProperties['zIndex'];

	/** top */
	top?: React.CSSProperties['top'];

	/** right */
	right?: React.CSSProperties['right'];

	/** bottom */
	bottom?: React.CSSProperties['bottom'];

	/** left */
	left?: React.CSSProperties['left'];

	/** box shadow */
	boxShadow?: React.CSSProperties['boxShadow'] | BoxShadowIntensityType;

	/** animation */
	animation?: React.CSSProperties['animation'];

	/** visibility */
	visibility?: React.CSSProperties['visibility'];

	/** cursor */
	cursor?: React.CSSProperties['cursor'];

	/** rounds the element */
	round?: IntensityType;

	/** white space */
	whiteSpace?: React.CSSProperties['whiteSpace'];

	/** hover */
	_hover?: string;

	/** all */
	all?: React.CSSProperties['all'];

	/** transition */
	transition?: React.CSSProperties['transition'];

	/** hide box based on device size */
	lgHide?: boolean;
	mdHide?: boolean;
	smHide?: boolean;
	xsHide?: boolean;
}
