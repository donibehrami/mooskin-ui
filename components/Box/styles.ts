import styled from 'styled-components';

// Screen sizes
import screens from '../_utils/globals/screens';

// Models
import { IStyledTheme } from '../Styled/model';
import { BoxShadowIntensityType, IBoxComponentProps, IntensityType, NestedThemeType } from './model';

export const StyledBox = styled.div<IBoxComponentProps>`
	&&& {
		margin: ${(props) => getNumberOrStringValue(props.m)};
		margin-top: ${(props) => getNumberOrStringValue(props.mt || props.my)};
		margin-right: ${(props) => getNumberOrStringValue(props.mr || props.mx)};
		margin-bottom: ${(props) => getNumberOrStringValue(props.mb || props.my)};
		margin-left: ${(props) => getNumberOrStringValue(props.ml || props.mx)};

		padding: ${(props) => getNumberOrStringValue(props.p)};
		padding-top: ${(props) => getNumberOrStringValue(props.pt || props.py)};
		padding-right: ${(props) => getNumberOrStringValue(props.pr || props.px)};
		padding-bottom: ${(props) => getNumberOrStringValue(props.pb || props.py)};
		padding-left: ${(props) => getNumberOrStringValue(props.pl || props.px)};

		color: ${(props) => getNestedValue(props.color, props.palette)};
		font-family: ${(props) => props.fontFamily};
		font-size: ${(props) => getNumberOrStringValue(props.fontSize)};
		font-weight: ${(props) => props.fontWeight};
		line-height: ${(props) => props.lineHeight};
		text-align: ${(props) => props.textAlign};
		font-style: ${(props) => props.fontStyle};
		text-transform: ${(props) => props.textTransform};
		text-decoration: ${(props) => props.textDecoration};
		white-space: ${(props) => props.whiteSpace};

		width: ${(props) => getNumberOrStringValue(props.w)};
		height: ${(props) => getNumberOrStringValue(props.h)};
		min-width: ${(props) => getNumberOrStringValue(props.minW)};
		max-width: ${(props) => getNumberOrStringValue(props.maxW)};
		min-height: ${(props) => getNumberOrStringValue(props.minH)};
		max-height: ${(props) => getNumberOrStringValue(props.maxH)};

		display: ${(props) => props.d};

		overflow: ${(props) => props.overflow};
		overflow-y: ${(props) => props.overflowY};
		overflow-x: ${(props) => props.overflowX};

		align-items: ${(props) => props.align};
		align-content: ${(props) => props.alignContent};
		align-self: ${(props) => props.alignSelf};
		justify-content: ${(props) => props.justify};
		justify-items: ${(props) => props.justifyItems};
		justify-self: ${(props) => props.justifySelf};
		flex-wrap: ${(props) => props.flexWrap};
		flex-direction: ${(props) => props.direction};
		flex: ${(props) => props.flex};
		flex-grow: ${(props) => props.flexGrow};
		flex-shrink: ${(props) => props.flexShrink};
		flex-basis: ${(props) => getNumberOrStringValue(props.flexBasis)};
		order: ${(props) => props.order};

		background: ${(props) => props.bg};
		background-image: ${(props) => props.bgImage};
		background-color: ${(props) => getNestedValue(props.bgColor, props.palette)};
		background-size: ${(props) => props.bgSize};
		background-position: ${(props) => props.bgPosition};
		background-repeat: ${(props) => props.bgRepeat};

		border: ${(props) => props.border};
		border-style: ${(props) => props.borderStyle};
		border-color: ${(props) => getNestedValue(props.borderColor, props.palette)};
		border-width: ${(props) => getNumberOrStringValue(props.borderWidth)};

		border-top: ${(props) => props.borderTop};
		border-top-style: ${(props) => props.borderTopStyle};
		border-top-color: ${(props) => props.borderTopColor};
		border-top-width: ${(props) => getNumberOrStringValue(props.borderTopWidth)};

		border-right: ${(props) => props.borderRight};
		border-right-style: ${(props) => props.borderRightStyle};
		border-right-color: ${(props) => props.borderRightColor};
		border-right-width: ${(props) => getNumberOrStringValue(props.borderRightWidth)};

		border-bottom: ${(props) => props.borderBottom};
		border-bottom-style: ${(props) => props.borderBottomStyle};
		border-bottom-color: ${(props) => props.borderBottomColor};
		border-bottom-width: ${(props) => getNumberOrStringValue(props.borderBottomWidth)};

		border-left: ${(props) => props.borderLeft};
		border-left-style: ${(props) => props.borderLeftStyle};
		border-left-color: ${(props) => props.borderLeftColor};
		border-left-width: ${(props) => getNumberOrStringValue(props.borderLeftWidth)};

		border-radius: ${(props) => (props.borderRadius ? getNumberOrStringValue(props.borderRadius) : getRoundness(props.round))};
		border-top-left-radius: ${(props) => getNumberOrStringValue(props.borderTopLeftRadius)};
		border-top-right-radius: ${(props) => getNumberOrStringValue(props.borderTopRightRadius)};
		border-bottom-right-radius: ${(props) => getNumberOrStringValue(props.borderBottomRightRadius)};
		border-bottom-left-radius: ${(props) => getNumberOrStringValue(props.borderBottomLeftRadius)};

		position: ${(props) => props.position};
		z-index: ${(props) => props.zIndex};
		top: ${(props) => getNumberOrStringValue(props.top)};
		right: ${(props) => getNumberOrStringValue(props.right)};
		bottom: ${(props) => getNumberOrStringValue(props.bottom)};
		left: ${(props) => getNumberOrStringValue(props.left)};

		box-shadow: ${(props) => getBoxShadow(props.boxShadow)};

		animation: ${(props) => props.animation};

		visibility: ${(props) => props.visibility};

		opacity: ${(props) => props.opacity};

		cursor: ${(props) => props.cursor};

		:hover {
			${(props) => props._hover}
		}

		${(props) => generateHiddenMediaQuery(props)}
	}
`;

StyledBox.displayName = 'StyledBox';

export const getNumberOrStringValue = (value?: number | string) => {
	if (typeof value !== 'undefined') {
		return typeof value === 'number' ? `${value}px` : value;
	}
};

export const getNestedValue = (value?: number | string | NestedThemeType, pallete?: IStyledTheme) => {
	if (Array.isArray(value) && pallete) {
		return (pallete as any)[value[0]][value[1]];
	}
	return value;
};

export const getRoundness = (round?: IntensityType) => {
	switch (round) {
		case 'xs':
			return '0.125rem';

		case 'sm':
			return '0.25rem';

		case 'base':
			return '0.375rem';

		case 'md':
			return '0.5rem';

		case 'lg':
			return '0.625rem';

		case 'xl':
			return '0.75rem';

		case '2xl':
			return '1rem';

		default:
			break;
	}
};

export const getBoxShadow = (boxShadow?: React.CSSProperties['boxShadow'] | BoxShadowIntensityType) => {
	switch (boxShadow) {
		case 'xs':
			return '0 0 0 1px rgba(0, 0, 0, 0.05)';

		case 'sm':
			return '0 1px 2px 0 rgba(0, 0, 0, 0.05)';

		case 'base':
			return '0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)';

		case 'md':
			return '0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)';

		case 'lg':
			return '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)';

		case 'xl':
			return '0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)';

		case '2xl':
			return '0 25px 50px -12px rgba(0, 0, 0, 0.25)';

		case 'outline':
			return '0 0 0 3px rgba(66, 153, 225, 0.6)';

		case 'inner':
			return 'inset 0 2px 4px 0 rgba(0,0,0,0.06)';

		default:
			return boxShadow;
	}
};

const generateHiddenMediaQuery = (props: IBoxComponentProps) => {
	const { lgHide, mdHide, smHide, xsHide } = props;
	let media = '';

	if (lgHide) {
		media = `${media} ${getQueryString(screens.large)}`;
	}

	if (mdHide) {
		media = `${media} ${getQueryString(screens.medium)}`;
	}

	if (smHide) {
		media = `${media} ${getQueryString(screens.small)}`;
	}

	if (xsHide) {
		media = `${media} ${getQueryString(screens.xSmall)}`;
	}

	return media;
};

const getQueryString = (media: string) => {
	return `
        @media ${media} {
            display: none;
        }
    `;
};
