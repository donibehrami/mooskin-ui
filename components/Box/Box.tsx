import * as React from 'react';

// Models
import { IDivBoxComponentProps } from './model';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Styled Components
import {StyledBox} from './styles';

export const Box: React.FC<IDivBoxComponentProps> = withMooskinContext((props) => {
    return <StyledBox {...props} as={props.boxAs} />;
});

Box.defaultProps = {
    className: '',
    style: {}
};

Box.displayName = 'Box';

export default Box;
