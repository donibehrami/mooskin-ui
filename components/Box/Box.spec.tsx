import * as React from 'react';
import {Box} from './Box';

import {shallow} from 'enzyme';

const boxStyle = {
    height: 40,
    width: 40
};

describe('Box', () => {

    test('renders Box correctly', () => {

        const tree = shallow(
            <Box>
                Box
            </Box>
        );

        expect(tree).toMatchSnapshot();
    });

});
