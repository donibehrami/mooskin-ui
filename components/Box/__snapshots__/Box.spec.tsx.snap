// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders Box correctly 1`] = `
<Box
  align="baseline"
  className=""
  m={5}
  p={5}
  style={Object {}}
>
  <Component
    align="baseline"
    className=""
    m={5}
    p={5}
    palette={
      Object {
        "backgroundColors": Object {
          "button": "#3fbaca",
          "common": "#ffffff",
          "disabledButton": "#e2e2e2",
          "main": "#f7f7f7",
          "secondaryToggle": "#ffffff",
          "toggle": "#425579",
        },
        "borderColors": Object {
          "common": "#e2e2e2",
        },
        "fontColors": Object {
          "anchor": "#3fbaca",
          "button": "#ffffff",
          "common": "#293346",
          "description": "#9d9d9d",
          "disabledAnchor": "#9d9d9d",
          "label": "#2d2d2d",
          "secondaryToggle": "#9d9d9d",
          "text": "#2d2d2d",
          "toggle": "#ffffff",
        },
      }
    }
    style={Object {}}
  >
    <StyledBox
      align="baseline"
      className=""
      m={5}
      p={5}
      palette={
        Object {
          "backgroundColors": Object {
            "button": "#3fbaca",
            "common": "#ffffff",
            "disabledButton": "#e2e2e2",
            "main": "#f7f7f7",
            "secondaryToggle": "#ffffff",
            "toggle": "#425579",
          },
          "borderColors": Object {
            "common": "#e2e2e2",
          },
          "fontColors": Object {
            "anchor": "#3fbaca",
            "button": "#ffffff",
            "common": "#293346",
            "description": "#9d9d9d",
            "disabledAnchor": "#9d9d9d",
            "label": "#2d2d2d",
            "secondaryToggle": "#9d9d9d",
            "text": "#2d2d2d",
            "toggle": "#ffffff",
          },
        }
      }
      style={Object {}}
    >
      <div
        className="sc-bdnylx iuGzom"
        style={Object {}}
      >
        Box
      </div>
    </StyledBox>
  </Component>
</Box>
`;
